test/integration/http_server_options_test.lua

```diff
@@ -0,0 +1,170 @@
+    g.httpd = helpers.cfgserv()
+    g.httpd:start()
+end)
+
+local IDLE_TIMEOUT = 1
+
+g.before_test('test_enabled_idle_timeout', function()
+    g.httpd = helpers.cfgserv({
+        idle_timeout = IDLE_TIMEOUT,
+        log_requests = true,
+        log_errors = true,
+    })
+    g.httpd:start()
+end)
+
+g.before_test('test_disabled_idle_timeout', function()
+    g.httpd = helpers.cfgserv({
+        log_requests = true,
+        log_errors = true,
+    })
+    g.httpd:start()
+end)
+
+
+g.after_each(function()
+    helpers.teardown(g.httpd)
+end)
+
+g.test_keepalive_default = function(g)
+    local httpd = g.httpd
+    t.assert_equals(httpd.options.keepalive_disable, {})
+end
+
+g.test_keepalive_allowed = function()
+    local r = http_client.request('GET', helpers.base_uri .. '/test', nil, {
+        headers = {
+            ['user-agent'] = 'Mozilla/4.0',
+            connection = 'keep-alive'
+        },
+    })
+    t.assert_equals(r.status, 200)
+    t.assert_equals(r.headers.connection, 'keep-alive')
+end
+
+g.test_keepalive_disallowed = function()
+    local r = http_client.request('GET', helpers.base_uri .. '/test', nil, {
+        headers = {
+            ['user-agent'] = g.useragent,
+            connection = 'keep-alive'
+        },
+    })
+    t.assert_equals(r.status, 200)
+    t.assert_equals(r.headers.connection, 'close')
+end
+
+g.test_keepalive_change_in_runtime = function(g)
+    local httpd = g.httpd
+    t.assert_equals(httpd.options.keepalive_disable, {})
+    httpd.options.keepalive_disable = {
+        'Mozilla/4.0',
+    }
+    t.assert_equals(httpd.options.keepalive_disable, {
+        'Mozilla/4.0',
+    })
+end
+
+g.test_idle_timeout_default = function(g)
+    local httpd = g.httpd
+    t.assert_equals(httpd.options.idle_timeout, TIMEOUT_INFINITY)
+    t.assert_equals(httpd.options.num_connections, nil)
+end
+
+g.test_enabled_idle_timeout = function(g)
+    local httpd = g.httpd
+    t.assert_equals(httpd.options.idle_timeout, IDLE_TIMEOUT)
+    t.assert_equals(httpd.num_connections, 0)
+
+    -- Open a keepalive connection on http server with idle_timeout equal to
+    -- IDLE_TIMEOUT. TCP server will start fiber with handler that will
+    -- increment num_connections and then will enter to busy loop.
+    local response = http_client.request('GET', helpers.base_uri .. '/test', nil, {
+        headers = {
+            connection = 'keep-alive'
+        },
+    })
+    -- Make sure parameters are correct.
+    t.assert_equals(response.status, 200)
+    t.assert_equals(response.headers.connection, 'keep-alive')
+    t.assert_equals(httpd.num_connections, 1)
+
+    -- Wait for IDLE_TIMEOUT, after that connection must be closed by timeout.
+    require('fiber').sleep(IDLE_TIMEOUT)
+
+    -- Server will handle http request using previously started fiber if
+    -- connection was not closed. If connection was closed then server will
+    -- start new fiber with handler that it will increment num_connections.
```

Comment:
how can we be sure, that a new connection is created, if we always check only for `num_connections == 1`?

Reply:
Using internal handlers.